core/src/main/java/forklift/concurrent/BlockingRejectedExecutionHandler.java

```diff
@@ -0,0 +1,27 @@
+package forklift.concurrent;
+
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.RejectedExecutionHandler;
+
+/**
+ * A rejected execution handler for a {@link ThreadPoolExecutor} that handles rejected
+ * tasks from a running thread pool by blocking.
+ */
+public class BlockingRejectedExecutionHandler implements RejectedExecutionHandler {
+    private static final Runnable NO_OP = () -> {};
+
+    @Override
+    public void rejectedExecution(Runnable runnable, ThreadPoolExecutor executor) {
+        if (executor.isShutdown()) { return; }
+
+        try {
+            // wait for space in the queue
+            executor.getQueue().put(NO_OP);
+            executor.remove(NO_OP);
```

Comment:
I'm not too familiar with RejectedExecutionHandlers but, why not add the runnable directly to the queue?  Is it to avoid the case where you add to the queue while the executor is shutting down?  Also, would it be quicker to just allow the NO_OP to run rather than remove it?

Reply:
>  Is it to avoid the case where you add to the queue while the executor is shutting down?

Yup. 

Since there's no way (to my knowledge) a generic queue itself can stop from receiving elements, using `put` you could get out-of-order execution doing something normally reasonable like:
```java
pool.shutdown(); // stop accepting tasks, set the pool to be on the path to being terminated
pool.getQueue().clear(); // remove any old tasks
/* probable context switch (since blocking is implemented with wait-notify) 
   -- waiting element gets queued */
pool.awaitTermination(...); // wait

pool.shutdownNow(); // interrupt any operations, there's no more time!
```

Basically, adding a harmless element is an easy way to wait for a `BlockingQueue` to have space. The other alternative is to do a poll loop on `BlockingQueue#remainingCapacity` without a wait-notify.

Realistically, this particular implementation isn't a huge concern since the rejected execution handler runs on the same thread, so a shutdown can't happen in the middle of the rejected execution handler. If you don't like it, I can just do `executor.getQueue().put(action)` and then throw it in a subclass of `Consumer` to hide it from misuse (after all, it could probably still be misused as-is by having multiple threads feed the same thread pool).

> Also, would it be quicker to just allow the NO_OP to run rather than remove it?

Huh? It should be at the back end of an almost full queue, so that seems like that would be equivalent to waiting for all of the elements in the initial queue (before waiting) to complete?